File: android/src/main/kotlin/xyz/bczl/xlog/XlogConfigApiImpl.kt
package xyz.bczl.xlog

/**
 * Native implementation of the pigeon-generated [XlogConfigApi].
 *
 * Bridges the Dart-side configuration calls to the [XLog] wrapper. This exists
 * so a pure-Flutter app can initialize xlog from Dart; a hybrid app may instead
 * call [XLog.init] directly from native code and never touch this API.
 *
 * Only configuration is handled here. Writing log entries goes through Dart FFI.
 */
class XlogConfigApiImpl : XlogConfigApi {

    override fun initialize(config: XlogConfig) {
        XLog.init(
            cacheDir = config.cacheDir,
            logDir = config.logDir,
            prefix = config.namePrefix,
            level = config.level.toLogLevel(),
            mode = config.mode.toAppenderMode(),
            pubKey = config.pubKey,
            compressMode = config.compressMode.toCompressMode(),
            compressLevel = config.compressLevel.toInt(),
            cacheDays = config.cacheDays.toInt(),
        )
        XLog.setConsoleLogEnabled(config.consoleLogEnabled)
    }

    override fun flush(sync: Boolean) {
        if (sync) XLog.flushSync() else XLog.flush()
    }

    override fun close() {
        XLog.close()
    }

    override fun setLevel(level: XlogLevel) {
        XLog.setLevel(level.toLogLevel())
    }

    override fun getLevel(): XlogLevel = XLog.getLevel().toXlogLevel()

    override fun setConsoleLogEnabled(enabled: Boolean) {
        XLog.setConsoleLogEnabled(enabled)
    }

    override fun setAppenderMode(mode: XlogMode) {
        XLog.setAppenderMode(mode.toAppenderMode())
    }

    override fun setMaxFileSize(maxBytes: Long) {
        XLog.setMaxFileSize(maxBytes)
    }

    override fun setMaxAliveSeconds(seconds: Long) {
        XLog.setMaxAliveTime(seconds)
    }

    // -------------------------------------------------------------------------
    // Multi-instance management (instances are created natively).
    // -------------------------------------------------------------------------

    override fun instancePrefixes(): List<String> = XLog.instancePrefixes()

    override fun hasInstance(prefix: String): Boolean =
        XLog.getLogInstance(prefix) != null

    override fun flushInstance(prefix: String, sync: Boolean) {
        val instance = XLog.getLogInstance(prefix) ?: return
        if (sync) instance.flushSync() else instance.flush()
    }

    override fun getInstanceLevel(prefix: String): XlogLevel =
        (XLog.getLogInstance(prefix)?.getLevel() ?: LogLevel.NONE).toXlogLevel()

    override fun setInstanceConsoleLogEnabled(prefix: String, enabled: Boolean) {
        XLog.getLogInstance(prefix)?.setConsoleLogEnabled(enabled)
    }

    override fun closeInstance(prefix: String) {
        XLog.closeInstance(prefix)
    }

    override fun getInstanceLogFiles(prefix: String, timespanDays: Long): List<String> =
        XLog.getLogInstance(prefix)?.getLogFiles(timespanDays.toInt()) ?: emptyList()
}

private fun XlogLevel.toLogLevel(): LogLevel = when (this) {
    XlogLevel.VERBOSE -> LogLevel.VERBOSE
    XlogLevel.DEBUG -> LogLevel.DEBUG
    XlogLevel.INFO -> LogLevel.INFO
    XlogLevel.WARNING -> LogLevel.WARNING
    XlogLevel.ERROR -> LogLevel.ERROR
    XlogLevel.FATAL -> LogLevel.FATAL
    XlogLevel.NONE -> LogLevel.NONE
}

private fun LogLevel.toXlogLevel(): XlogLevel = when (this) {
    LogLevel.VERBOSE -> XlogLevel.VERBOSE
    LogLevel.DEBUG -> XlogLevel.DEBUG
    LogLevel.INFO -> XlogLevel.INFO
    LogLevel.WARNING -> XlogLevel.WARNING
    LogLevel.ERROR -> XlogLevel.ERROR
    LogLevel.FATAL -> XlogLevel.FATAL
    LogLevel.NONE -> XlogLevel.NONE
}

private fun XlogMode.toAppenderMode(): Int = when (this) {
    XlogMode.ASYNC -> XLog.MODE_ASYNC
    XlogMode.SYNC -> XLog.MODE_SYNC
}

private fun XlogCompressMode.toCompressMode(): Int = when (this) {
    XlogCompressMode.ZLIB -> XLog.COMPRESS_ZLIB
    XlogCompressMode.ZSTD -> XLog.COMPRESS_ZSTD
}

File: ios/Classes/XlogConfigApiImpl.swift
import Foundation

/// Bridges Flutter configuration calls to the same facade available to native code.
class XlogConfigApiImpl: XlogConfigApi {
  func initialize(config: XlogConfig) throws {
    let configuration = XLogConfiguration(
      logDirectory: config.logDir,
      namePrefix: config.namePrefix
    )
    configuration.cacheDirectory = config.cacheDir.isEmpty ? nil : config.cacheDir
    configuration.publicKey = config.pubKey.isEmpty ? nil : config.pubKey
    configuration.cacheDays = Int(config.cacheDays)
    configuration.level = config.level.logLevel
    configuration.mode = config.mode.xlogMode
    configuration.compressMode = config.compressMode.xlogCompressMode
    configuration.compressLevel = Int(config.compressLevel)
    configuration.consoleLogEnabled = config.consoleLogEnabled

    guard XLog.initialize(configuration: configuration) else {
      throw PigeonError(
        code: "initialize_failed",
        message: "mars-xlog failed to open the global appender",
        details: nil
      )
    }
  }

  func flush(sync: Bool) throws {
    sync ? XLog.flushSync() : XLog.flush()
  }

  func close() throws {
    XLog.close()
  }

  func setLevel(level: XlogLevel) throws {
    XLog.level = level.logLevel
  }

  func getLevel() throws -> XlogLevel {
    XLog.level.pigeonLevel
  }

  func setConsoleLogEnabled(enabled: Bool) throws {
    XLog.setConsoleLogEnabled(enabled)
  }

  func setAppenderMode(mode: XlogMode) throws {
    XLog.setAppenderMode(mode.xlogMode)
  }

  func setMaxFileSize(maxBytes: Int64) throws {
    XLog.setMaxFileSize(maxBytes)
  }

  func setMaxAliveSeconds(seconds: Int64) throws {
    XLog.setMaxAliveTime(TimeInterval(seconds))
  }

  func instancePrefixes() throws -> [String] {
    XLog.instanceNamePrefixes
  }

  func hasInstance(prefix: String) throws -> Bool {
    XLog.instance(namePrefix: prefix) != nil
  }

  func flushInstance(prefix: String, sync: Bool) throws {
    guard let instance = XLog.instance(namePrefix: prefix) else { return }
    sync ? instance.flushSync() : instance.flush()
  }

  func getInstanceLevel(prefix: String) throws -> XlogLevel {
    (XLog.instance(namePrefix: prefix)?.level ?? .none).pigeonLevel
  }

  func setInstanceConsoleLogEnabled(prefix: String, enabled: Bool) throws {
    XLog.instance(namePrefix: prefix)?.setConsoleLogEnabled(enabled)
  }

  func closeInstance(prefix: String) throws {
    XLog.closeInstance(namePrefix: prefix)
  }

  func getInstanceLogFiles(prefix: String, timespanDays: Int64) throws -> [String] {
    XLog.instance(namePrefix: prefix)?.logFiles(timeSpanDays: Int(timespanDays)) ?? []
  }
}

private extension XlogLevel {
  var logLevel: LogLevel {
    LogLevel(rawValue: rawValue) ?? .none
  }
}

private extension LogLevel {
  var pigeonLevel: XlogLevel {
    XlogLevel(rawValue: rawValue) ?? .none
  }
}

private extension XlogMode {
  var xlogMode: XLogMode {
    XLogMode(rawValue: rawValue) ?? .async
  }
}

private extension XlogCompressMode {
  var xlogCompressMode: XLogCompressMode {
    XLogCompressMode(rawValue: rawValue) ?? .zlib
  }
}
